design/README.md

```diff
@@ -0,0 +1,73 @@
+# Code specification
+
+## Motivation
+
+Replace the current APFELcomb code with a simpler and faster tool to generate FK Tables for NNPDF fits.
+
+## APFELcomb design and issues
+
+### Workflow
+
+The apfelcomb project compiles a program which connects several tools and
+databases from NNPDF and external providers in order to generate FK Tables,
+where observables are stored into tables with a predefined input scale energy value for PDFs evaluation.
+
+The user/develop should include the custom implementation of the combination operation for each specific process category. All operations involving DGLAP, DIS and Fixed-target DY are performed at runtime. Operations involving the generation of hadronic observables with or without electroweak corrections are performed off-line and stored on git-lfs repositories.
+
+The mapping between the requested FK table configuration, in terms of theory and dataset choice, are controlled by the NNPDF theory sqlite database and by a custom sqlite database build inside apfelcomb.
+
+### Inputs
+
+The apfelcomb program takes as input the:
+- the theory database from NNPDF,
+- a custom database with specifications for each dataset in NNPDF,
+- CommonData files from NNPDF/buildmaster,
+- APPLgrid grids,
+- PineAPPL grids.
+
+### External tools
+
+The current apfelcomb implementation relies on third-party tools for:
+- DGLAP evolution kernels (APFEL),
+- DIS observables (APFEL),
+- Fixed-target DY (APFEL)
+- APPLgrids for hadronic processes (APPLgrid),
+- PineAPPL grids for processes with EW corrections (PineAPPL).
+
+### Problems to be avoided
+
+Here a short summary of practical problems we would like to avoid:
+- avoid different data input format, today we have to maintain parsers for APFEL (DIS/FTDY), APPLgrid and PineAPPL,
+- avoid long runtime operations that can be cached and stored in a server a priori, e.g.: DGLAP evolution kernels, DIS and FTDY grids,
+- reduce to minimum data input processing,
+- delegate the FK Table construction to a single code/function,
+- provide tools to cache results and upload to server
+- provide tools to optimize the quality of FK tables.
```

Comment:
Here are more problems that we can avoid:

- after generation we want to check three numbers with a realistic PDF (a toy PDF is not enough!): 1) the PineAPPL predictions, 2) the predictions using EKO with a reduced number of grid points (measures the compression penalty), and 3) the prediction of the FK table (measures the effect of the down-evolution). The last two numbers are the ones shown already in apfelcomb, but we really need the first number as well, as seen with ATLAS Z high-mass DY
- personally, I don't like the format of apfelcomb's database at all: 1) why is there an ID for each grid? I'd rather invoke the program with the grid file and the theory identifier, 2) the ppbar switch should no longer be needed as this something that I'll implement in PineAPPL (see https://github.com/N3PDF/pineappl/issues/39), 3) the bin selection is very fragile - right now there's a '1' for every bin that's included (imagine doing that for ATLAS 3D DY with 504 bins), but we probably should do it the other around, by specifying cuts, 4) merge all the databases (DIS/FTDY/...) into a single one

Reply:
I fully agree with the first point. Concerning the second point, I think the main problem is that we have been using applgrids from multiple providers, with lack of quality control. The db is just reshaping the grids to match the NNPDF specs, in terms of datapoints, units, masks, normalization, etc. With pineappl we can get rid of all that and create a 1:1 match between the dataset and the theory grid.

Reply:
Exactly; I think to summarise my second point, what we should avoid is to put the relevant information in too many places - this makes all the procedures very error-prone and fragile, as we all know.